src/java.base/share/classes/java/util/regex/MatchResult.java

```diff
@@ -184,6 +307,49 @@ public interface MatchResult {
      *
      * @return The number of capturing groups in this matcher's pattern
      */
-    public int groupCount();
+    int groupCount();
+
+    /**
+     * Returns an unmodifiable map from capturing group names to group numbers.
+     * If there are no named groups, returns an empty map.
+     *
+     * @return an unmodifiable map from capturing group names to group numbers
+     *
+     * @throws UnsupportedOperationException
+     *          The default implementation of this method always throws
+     *
+     * @apiNote
+     * This method must be overridden by an implementation that supports
+     * named groups.
+     *
+     * @since 20
+     */
+    default Map<String,Integer> namedGroups() {
+        throw new UnsupportedOperationException("namedGroups()");
+    }
+
+    private int groupNumber(String name) {
+        Objects.requireNonNull(name, "Group name");
+        Integer number = namedGroups().get(name);
+        if (number != null) {
+            return number;
+        }
+        throw new IllegalArgumentException("No group with name <" + name + ">");
+    }
+
+    /**
+     * Returns whether {@code this} contains a valid match from
+     * a previous match or find operation.
+     *
+     * @return whether {@code this} contains a valid match
+     *
+     * @throws UnsupportedOperationException
+     *          The default implementation of this method always throws
```

Comment:
Similar comment here as above, though it has nothing to do with named groups. The `@throws` clause should say it throws UOE if "the implementation cannot report whether or not it has a match" or some such. This is a bit odd, but the specification needs to be permissive enough so that it doesn't invalidate existing implementations outside the JDK.

As before, the "always throws" should be moved to `@implSpec`.

Reply:
Done